### ISSUE TITLE
Does anyone have a guide for getting this running on an AWS GPU instance?

### ISSUE THREAD

**Issue by spullara** [kingoflolz/mesh-transformer-jax#17]
I think I've got everything installed but it is failing. It tries and fails to use TPU rather than GPU.

**Comment by kingoflolz** [comment]
Did you follow https://github.com/google/jax#installation to install JAX with CUDA? I'm not able to help much with JAX installation/setup.

**Comment by spullara** [comment]
I did. I think some stuff has later versions though. I will try it from scratch again. For example, jax.experimental.maps has changed somewhat since the code was written.

**Comment by kingoflolz** [comment]
yeah, you need to use a specific jax version (0.2.12)

**Comment by paulcjh** [comment]
Any luck on this? I tried to run the model on a single T4 and I believe it's just overflowing when reading the checkpoint, makes sense based on the memory required.
Anyone had any luck running on multiple GPUs? I'm trying now to run on a g4dn.12xlarge but getting this error (changed the cores_per_replica to 4, and the error below is arrising when defining `CausalTransformer(params)`):
`RuntimeError: Internal: external/org_tensorflow/tensorflow/compiler/xla/service/gpu/nccl_utils.cc:198: NCCL operation ncclGroupEnd() failed: unhandled system error: while running replica 0 and partition 0 of a replicated computation (other replicas may have failed as well).`

**Comment by oborchers** [comment]
@maintainers. Do think it will be practical to have a torch/huggingface version for this in the future?

**Comment by kingoflolz** [comment]
@paulcjh The current resharding code does not work for > 1 output shard, and has only been tested on GPUs with more VRAM than a T4.

@oborchers There is a HF/torch port here https://github.com/huggingface/transformers/pull/12106 (@paulcjh you might want to try this too, people have had success running it on a 16GB card)

**Comment by paulcjh** [comment]
@kingoflolz yeah I've been following that now thanks. I have an Nvidia K80 (24GB) + Quadro RTX 8000 (48GB) coming today and tomorrow, I'll post results of the testing in here if they work well.

**Comment by oborchers** [comment]
@kingoflolz and @paulcjh: Excellent! I am really looking forward to this!! If you have a benchmark script available I can run it on a V100 as well for comparison.

**Comment by paulcjh** [comment]
Got everything running here on AWS in production, all on T4 instances https://hub.getneuro.ai/model/nlp/gpt-j-6B-text-generation

**Comment by spullara** [comment]
@paulcjh very cool. do you have a script I could look at to see what I am missing?

**Comment by paulcjh** [comment]
@spullara sure, got some of my work here: https://github.com/paulcjh/gpt-j-6b/blob/main/gpt-j-t4.ipynb

**Comment by oborchers** [comment]
@paulcjh : Thats rad! Thank you very much! Am able to run it that wa!!

**Comment by heisen273** [comment]
@paulcjh, @oborchers  since you guys tested it - perhaps you can let me know what was the model performance during your tests & GPUs(i.e. what was the time_taken for generating corpus based on string `"EleutherAI is"`) that you've used?

I'm considering whether it's a good idea to upgrade my videocard to use the GPT-j model.

Thank you very much for letting me know.

**Comment by oborchers** [comment]
@heisen273 sure! Tests on a V100 in half precision mode with the experimental gpt-j branch. Idle load ~12G, workload ~15G. Average watts consumed: ~130W. Utilization: 50%.

Params:
```
"do_sample": True,
"min_length": 1,
"max_length": 1024,
"early_stopping": True,
"repetition_penalty": 1.0,
"num_beams": 1,
"eos_token_id": 198, # "\n"
```

Benchmark:
```
generating 50 sequences took 125.29s or 2.51s/output. average num characters: 254.72
```

Example output:

```
... a team-oriented player known for his excellent team fighting, excellent knowledge of the game mechanics and his ability to contribute by clearing the creep waves for his team (and his ability to clean out a tower himself). This makes him a very valuable asset for any team willing to have him on their side.
```

**Comment by spullara** [comment]
All good for me. @paulcjh's solution is working on AWS GPU instances.

**Comment by dimitry12** [comment]
@paulcjh did it work on your K80? I assume not, because it full precision it would be 18GB and K80 is 12+12 - but want to double-check. Thanks!

**Comment by Pathos14489** [comment]
> 
> 
> @paulcjh did it work on your K80? I assume not, because in full precision it would be 18GB and K80 is 12+12 - but want to double-check. Thanks!

I have a K80 and would be willing to try running it, but I'm not entirely sure how.

**Comment by trisimix** [comment]
> 

did anyone figure this out? they are really cheap

**Comment by dimitry12** [comment]
I couldn't make k80 work in my threadripper system: either GPU was a dud,
or it wasn't compatible with my motherboard (common problem)

Trying to obtain rtx 3090

**Comment by trisimix** [comment]
Ah rip, thanks for sharing though.
